content/blog/2019-11-21-bug-finding/index.md

```diff
@@ -0,0 +1,308 @@
+
+1. the widths and representations of integers
+2. behavior when casting to a signed integer type when the value cannot be represented in an object of the target type
+3. results of bitwise operations on signed integers. 
+
+There are roughly three kinds of compiler targets:
+
+| Targets               | `int` bit length | `long` bit length |
+| --------------------- | ---------------- | ----------------- |
+| x86-64                | 32               | 64                |
+| x86, ARM, and PowerPC | 32               | 32                |
+| MSP430 and AVR        | 16               | 16                |
+
+Csmith performs testing within but not cross the equivalent classes.
+
+### No Ground Truth
+
+Csmith does not have a ground truth since it is unrealistic to have a human check each program. Instead, it takes the majority vote. In practice, two uncorrelated compilers have not output the same incorrect output, according to the authors. The explanation to this is the substantial diversity among intermediate languages (IRs).
+
+### No Guarantee of Termination
+
+A Csmith-generated program can be of any length. In practice, a time-out function is called to terminate program that takes too long to finish.
+
+### Target Middle-End bugs
+
+Csmith targets checking how compilers perform transformations on IRs rather than standard conformance as commercial test suites. For instance, Csmith does not spend efforts to test how compilers handle long identifier names.
+
+There are several choices made because of the target:
+
+1. The Csmith designers manually tune the 80 probability variables to generate programs with a balanced mix of arithmetic and bitwise operations, of loops and straight-line code, and of single-level and multi-level indirections, *etc*.
+2. Encouraging Csmith to generate idiomatic code, e.g. loops that access all elements of an array.
+3. Discouraging Csmith from generating source-level diversity that is unlikely to improve the “coverage” of a compiler’s IR, e.g. additional levels of parentheses around expressions.
+4. Designing Csmith efficiently generates runnable programs of a few tens of thousands of lines long in a few seconds.
+
+## Evaluation
+
+### Opportunistic Bug Finding
+
+The Csmith designers tested 11 compilers and reports the errors to the developers. The commercial compiler developers do not care, while GCC and LLVM team response quickly. By the time the paper is formalized, 79 GCC bugs and 202 LLVM bugs (2% of all LLVM bug reports) are reported and most of them are fixed. CompCert is such a good compiler that the under-development version of CompCert is the only compiler Csmith cannot find wrong-code errors after 6 CPU-years of testing.
```

Comment:
But they did find bugs in CompCert originally…

Reply:
But they only found 2, I will add to the post.